[JSC] Remove use of Committed / Primordial state of AccessCase and cloning cases only when it is Generated and having CallLinkInfo

Constellation opened the pull request:
#### 103f05d044f0b05b0497833dba83a642526c229a
<pre>
[JSC] Remove use of Committed / Primordial state of AccessCase and cloning cases only when it is Generated and having CallLinkInfo
<a href="https://bugs.webkit.org/show_bug.cgi?id=273276">https://bugs.webkit.org/show_bug.cgi?id=273276</a>
<a href="https://rdar.apple.com/127076206">rdar://127076206</a>

Reviewed by Keith Miller.

This is a change towards making AccessCase just a constant feedback information instead of having mutating state.
This patch removes effective use of Committed / Primordial state since we do not care about them. Right now CallLinkInfo is tied to
AccessCase so we still need to know about Generated state and we need to clone it. But we don&apos;t need to clone AccessCase when it doesn&apos;t
have CallLinkInfo inside it. So this patch removes this cloning for AccessCases not holding CallLinkInfo*.

This is paving a path towards <a href="https://bugs.webkit.org/show_bug.cgi?id=156456.">https://bugs.webkit.org/show_bug.cgi?id=156456.</a> Once CallLinkInfo* gets wiped from AccessCase, we can completely
remove (1) Generated state since AccessCase now just becomes immutable data and (2) cloning feature since we no longer need this.

* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::commit):
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::doesJSCalls):
(JSC::InlineCacheCompiler::generateWithGuard):
(JSC::InlineCacheCompiler::generate):
(JSC::InlineCacheCompiler::regenerate):

Canonical link: <a href="https://commits.webkit.org/278051@main">https://commits.webkit.org/278051@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/518aeaca5604c3bdf4b3d741b4d8f676a57ad876

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49261 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28542 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52292 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52057 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/45364 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34556 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26163 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40238 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51362 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26087 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42435 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21361 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23540 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43611 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7583 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/42563 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45463 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44118 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53968 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/48750 "Build was cancelled. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled JSC (warnings); jscore-tests (cancelled)") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24342 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20529 "Found 1 new test failure: imported/w3c/web-platform-tests/html/canvas/offscreen/manual/filter/offscreencanvas.filter.w.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47558 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25615 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42637 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46560 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26453 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/56244 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7087 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25337 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11552 "Found 4 new JSC stress test failures: microbenchmarks/delete-property-keeps-cacheable-structure.js.dfg-eager, microbenchmarks/delete-property-keeps-cacheable-structure.js.dfg-eager-no-cjit-validate, microbenchmarks/getter-richards.js.default, microbenchmarks/getter-richards.js.no-cjit (failure)") | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
```diff
@@ -4343,7 +4441,7 @@ AccessGenerationResult InlineCacheCompiler::regenerate(const GCSafeConcurrentJSL
     m_doesCalls = doesCalls;
     m_doesJSCalls = doesJSCalls;
-    if (needsSymbolPropertyCheck || needsStringPropertyCheck || needsInt32PropertyCheck)
+    if (needsSymbolPropertyCheck || needsStringPropertyCheck || needsInt32PropertyCheck || doesJSCalls)
```
Nit add the comment from above: `Cases that do JS calls rely on stack-pointer adjustment, which is tied to the linked CodeBlock, which makes this code unshareable.`

Constellation replied:
In fact, this comment was stale and no longer true :) We are storing this information in JITData and we can adjust it without using CodeBlock embedded pointer (we are already doing that for Baseline JIT IC).
The current disablement is largely because CallLinkInfo pointer is embedded into the code itself. (And AccessCase holds it).

kmiller68 reviewed (approved):
r=me with nit.

Constellation committed:
[JSC] Remove use of Committed / Primordial state of AccessCase and cloning cases only when it is Generated and having CallLinkInfo
https://bugs.webkit.org/show_bug.cgi?id=273276
rdar://127076206

Reviewed by Keith Miller.

This is a change towards making AccessCase just a constant feedback information instead of having mutating state.
This patch removes effective use of Committed / Primordial state since we do not care about them. Right now CallLinkInfo is tied to
AccessCase so we still need to know about Generated state and we need to clone it. But we don't need to clone AccessCase when it doesn't
have CallLinkInfo inside it. So this patch removes this cloning for AccessCases not holding CallLinkInfo*.

This is paving a path towards https://bugs.webkit.org/show_bug.cgi?id=156456. Once CallLinkInfo* gets wiped from AccessCase, we can completely
remove (1) Generated state since AccessCase now just becomes immutable data and (2) cloning feature since we no longer need this.

* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::commit):
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::doesJSCalls):
(JSC::InlineCacheCompiler::generateWithGuard):
(JSC::InlineCacheCompiler::generate):
(JSC::InlineCacheCompiler::regenerate):

Canonical link: https://commits.webkit.org/278051@main
